File: app/src/main/java/com/worldventures/dreamtrips/social/ui/membership/service/command/SendInvitesCommand.java
package com.worldventures.dreamtrips.social.ui.membership.service.command;

import com.worldventures.core.janet.CommandWithError;
import com.worldventures.core.janet.dagger.InjectableAction;
import com.worldventures.dreamtrips.R;
import com.worldventures.dreamtrips.api.invitation.CreateInvitationHttpAction;
import com.worldventures.dreamtrips.api.invitation.model.CreateInvitationParams;
import com.worldventures.dreamtrips.api.invitation.model.ImmutableCreateInvitationParams;
import com.worldventures.dreamtrips.api.invitation.model.InvitationType;
import com.worldventures.dreamtrips.social.ui.membership.model.InviteTemplate;

import java.util.List;

import javax.inject.Inject;

import io.techery.janet.Janet;
import io.techery.janet.command.annotations.CommandAction;

@CommandAction
public class SendInvitesCommand extends CommandWithError implements InjectableAction {

   @Inject Janet janet;

   private int templateId;
   private List<String> contacts;
   private InviteTemplate.Type type;

   public SendInvitesCommand(int templateId, List<String> contacts, InviteTemplate.Type type) {
      this.templateId = templateId;
      this.contacts = contacts;
      this.type = type;
   }

   @Override
   public int getFallbackErrorMessage() {
      return R.string.error_fail_to_get_preview_for_template;
   }

   @Override
   protected void run(CommandCallback callback) throws Throwable {
      janet.createPipe(CreateInvitationHttpAction.class)
            .createObservableResult(new CreateInvitationHttpAction(provideParams()))
            .subscribe(callback::onSuccess, callback::onFail);
   }

   private CreateInvitationParams provideParams() {
      return ImmutableCreateInvitationParams.builder()
            .templateId(templateId)
            .contacts(contacts)
            .type(type == InviteTemplate.Type.SMS ? InvitationType.SMS : InvitationType.EMAIL)
            .build();
   }
}

File: app/src/main/java/com/worldventures/dreamtrips/social/service/invites/SendInvitesCommand.kt
package com.worldventures.dreamtrips.social.service.invites

import com.worldventures.core.janet.CommandWithError
import com.worldventures.dreamtrips.R
import com.worldventures.dreamtrips.api.invitation.CreateInvitationHttpAction
import com.worldventures.dreamtrips.api.invitation.model.ImmutableCreateInvitationParams
import com.worldventures.dreamtrips.api.invitation.model.InvitationType
import com.worldventures.dreamtrips.social.domain.entity.InviteType
import com.worldventures.janet.injection.InjectableAction
import io.techery.janet.Command
import io.techery.janet.Janet
import io.techery.janet.command.annotations.CommandAction
import javax.inject.Inject

@CommandAction
class SendInvitesCommand(private val templateId: Int, private val contacts: List<String>, private val type: InviteType)
   : CommandWithError<Boolean>(), InjectableAction {

   @field:Inject lateinit var janet: Janet

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<Boolean>) {
      janet.createPipe(CreateInvitationHttpAction::class.java)
            .createObservableResult(CreateInvitationHttpAction(provideParams()))
            .map { true }
            .subscribe(callback::onSuccess, callback::onFail)
   }

   private fun provideParams() = ImmutableCreateInvitationParams.builder()
         .templateId(templateId)
         .contacts(contacts)
         .type(if (type === InviteType.SMS) InvitationType.SMS else InvitationType.EMAIL)
         .build()

   override fun getFallbackErrorMessage() = R.string.error_fail_to_get_preview_for_template
}
